### PR TITLE
Support auto-running functions after set,get,update

### PR DESCRIPTION
**Pull request by pwittrock**
This allows a package to declare that some or all functions should be run after programmatically modifying the package through get,set,update.

### CODE REVIEW

File: internal/util/functions/functions_test.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+
+	"github.com/GoogleContainerTools/kpt/internal/util/functions"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestReconcileFunctions(t *testing.T) {
```
Comment (phanimarupaka) [inline internal/util/functions/functions_test.go]: Shouldn't we convert this to table tests.

File: internal/util/functions/functions_test.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        # head comment
+        image: nginx:1.8.1 # {"$ref": "#/definitions/io.k8s.cli.substitutions.image"}
+`
+
+	if !assert.Equal(t, expected, string(actual)) {
+		t.FailNow()
+	}
+}
+
+func TestReconcileFunctionsImplicit(t *testing.T) {
```
Comment (phanimarupaka) [inline internal/util/functions/functions_test.go]: Might be trivial, but shouldn't we add a case for autoRunContainer: true

File: internal/util/functions/functions_test.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	deploy := fmt.Sprintf(`apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  annotations:
+    config.kubernetes.io/function: |
```
Comment (phanimarupaka) [inline internal/util/functions/functions_test.go]: Does this mean this function must be applied only to this yaml file in the directory? Or is it applied for all the files. If yes, can we have one more file written to this directory and verify if it is applied or is it trivial?

File: internal/util/functions/functions_test.go (hunk 2 above)
Comment (pwittrock) [reply internal/util/functions/functions_test.go]: Testing running functions isn't easy due to the DiD problem.  I could delete the option from this PR.

File: internal/util/functions/functions_test.go (hunk 1 above)
Comment (pwittrock) [reply internal/util/functions/functions_test.go]: Done

File: internal/util/functions/functions_test.go (hunk 3 above)
Comment (pwittrock) [reply internal/util/functions/functions_test.go]: Done

**Comment by pwittrock** [comment]
PTAL

**Pull request merged**